### ISSUE TITLE
perf(whisper): decode loop 4x regression — missing async pipeline overlap (11ms vs 3ms per token)

### ISSUE THREAD

**Issue by SRjoeee** [findit-studio/mlxrs#362]
# Whisper decode loop: 4x performance regression (11ms vs 3ms per token)

## Root Cause

PR #320 (GPU logit filters) correctly moved logit filters to GPU, but introduced **2 synchronous `.item()` GPU→CPU round-trips per decode token** instead of Python's 1 async eval.

## Evidence

Component breakdown (whisper-base, 25s Chinese audio, M1 8GB):

| Metric | mlxrs Rust | mlx-audio Python | Ratio |
|--------|-----------|-----------------|-------|
| Encoder/window | 103ms | ~100ms | 1.0x ✅ |
| Per-token decode | **11.24ms** | ~3ms | **3.7x** ⛔ |
| Total RTF | 9.2x | 36.0x | 3.9x ⛔ |

### Before #314 (initial port): CPU path, no GPU sync
```rust
fn update(&mut self, logits: &[f32], last_token: u32) -> Result<(u32, bool)> {
    let next = argmax_slice(logits);           // CPU loop
    let lse = logsumexp_slice(logits);         // CPU loop
    let chosen_logprob = logits[next] - lse;   // CPU scalar
}
```

### After #320 (GPU logit filters): 2 GPU sync points per token
```rust
fn update(&mut self, logits: &Array, last_token: u32) -> Result<(u32, bool)> {
    let mut idx = ops::misc::argmax(logits, Some(-1), false)?;
    idx.item::<u32>()?  // SYNC #1: blocks GPU→CPU (~3-5ms)

    if last_token != self.eot {
        let chosen_logprob = self.chosen_logprob(logits, next)?;
        // chosen_logprob does:
        //   let lse = ops::reduction::logsumexp(logits, true)?;
        //   let mut logprob = ops::arithmetic::subtract(&chosen, &lse)?;
        //   Ok(logprob.item::<f32>()? as f64)  // SYNC #2: blocks GPU→CPU (~3-5ms)
        self.sum_logprob += chosen_logprob;
    }
}
```

### Python (mlx-audio): 1 async eval per token
```python
def update(self, logits, last_token):
    next_tokens = logits.argmax(axis=-1)           # LAZY
    logprobs = logits - mx.logsumexp(logits, -1)   # LAZY
    current_logprobs = logprobs[arange, next_tokens] # LAZY
    # ...
    mx.async_eval(next_completed, next_tokens, next_sum_logprobs)  # 1 eval
```

### Priority 1: Merge argmax + logprob into single graph (saves ~5-6ms)

```rust
fn update(&mut self, logits: &Array, last_token: u32) -> Result<(u32, bool)> {
    // Build the FULL lazy graph before any .item()
    let idx = ops::misc::argmax(logits, Some(-1), false)?;
    let lse = ops::reduction::logsumexp(logits, true)?;
    let chosen_idx = idx.item::<u32>()? as usize;  // 1 sync: evaluates BOTH argmax + lse
    // Now extract the chosen logprob from the already-materialized lse
    // ...
}
```

Or better: compute `logprob[argmax]` entirely lazily, then `.item()` once to read the scalar.

### Priority 2: Use async_eval in decode loop (saves ~2-3ms)

After selecting the token, call `async_eval` on remaining arrays so GPU work overlaps with CPU token management.

### Priority 3: Pre-allocate timestamp mask buffer (saves ~0.5-1ms)

`ApplyTimestampRules::deterministic_mask()` allocates `vec![0.0_f32; 51865]` (~200KB) every token. Pre-allocate once and reuse.

## Expected Result

After fixes: per-token ~3-4ms, total RTF ~30-36x (matching Python).

## Related PRs

- #320 — introduced the regression (GPU logit filters)
- #339, #345, #358 — decode loop unchanged, not the cause

**Comment by SRjoeee** [comment]
## Adversarial Validation Update (3 rounds)

The original analysis was imprecise. "2x .item() → 4x slowdown" framing is misleading. Here's the validated, corrected root cause:

### Corrected Root Cause: Missing async pipeline overlap

**Python** (decoding.py:616,625):
```python
# GreedyDecoder.update — ENTIRELY lazy, ZERO .item() calls
next_tokens = logits.argmax(axis=-1)           # lazy
logprobs = logits - mx.logsumexp(logits, -1)   # lazy
# ...
mx.async_eval(next_completed, next_tokens, next_sum_logprobs)  # GPU runs in background
# CPU immediately starts building next step's graph while GPU works
```

**Rust** (decoding.rs:657-686):
```rust
let mut idx = ops::misc::argmax(logits, Some(-1), false)?;
idx.item::<u32>()?  // BLOCKS: forces full decoder forward to complete (~90% of sync cost)
// GPU IDLE while CPU processes token
let chosen_logprob = self.chosen_logprob(logits, next)?;
// chosen_logprob: logsumexp + slice + subtract + .item::<f32>()?  (~10% of sync cost)
// GPU IDLE again
```

The issue is NOT the count of `.item()` calls. It's the **absence of async pipeline overlap**: Python's `async_eval` lets the GPU work on step N while the CPU builds step N+1's graph. Rust's synchronous `.item()` blocks until the entire decoder forward completes, leaving the GPU idle between iterations.

### Adversarial Test Results

| Test | Question | Result |
|------|----------|--------|
| 1 | Does Python use `async_eval`? | **YES** — lines 616, 625 confirmed |
| 2 | How much does each `.item()` cost? | First (argmax) ~90%, second (logsumexp) ~10% |
| 3 | Extra work in Rust decoder forward? | Redundant `astype(F32)` + `contiguous()` (~0.1ms) |
| 4 | Are logit filters slow? | `overwrite_masked` = 3 ops vs Python's `add` = 1 op, but lazy eval masks this |
| 5 | Safetensors format issue? | No impact |
| 6 | Remove chosen_logprob entirely? | Saves ~0.5-1ms, not root cause |

### Revised Fix Priority

1. **Implement async pipeline overlap** (saves ~7-8ms/token): Use `mlxrs::transforms::async_eval` or equivalent to overlap GPU compute with CPU token management. This is THE fix.
2. **Merge argmax + logsumexp into single lazy graph** (saves ~0.5-1ms): Compute both before any `.item()`, so one sync evaluates both.
3. **Remove redundant `astype(F32)` + `contiguous()`** in `last_position_row` (saves ~0.1ms).
4. **Pre-allocate `ApplyTimestampRules` mask buffer** (saves ~0.5-1ms per token).

Expected result: per-token ~3-4ms, matching Python's 36x realtime.

**Comment by SRjoeee** [comment]
## Third-Party Verification: Claude Code Independent Confirmation

Claude Code (sonnet) performed an independent pixel-level comparison of the mlxrs Rust decode loop vs mlx-audio Python decode loop. **Hypothesis fully confirmed.**

### Claude Code's Findings

**Python async_eval** — `decoding.py:616,625`:
```python
mx.async_eval(completed, tokens, sum_logprobs, no_speech_probs)   # line 616
mx.async_eval(next_completed, next_tokens, next_sum_logprobs)     # line 625
```
Pipeline overlap: GPU executes step N while CPU builds step N+1's graph. The `if completed:` check at line 626 reads the **previous** step's result (already materialized by async_eval).

**Rust synchronous .item()** — `decoding.rs:661-663, 698-700`:
```rust
idx.item::<u32>()?          // line 663 — SYNC #1: full decoder forward must complete
logprob.item::<f32>()?      // line 700 — SYNC #2: logsumexp + subtract must complete
```
Zero pipeline overlap. Every token serializes: GPU executes → CPU reads → GPU starts next.

### Additional Finding from Claude Code

**sum_logprobs accumulation strategy differs:**
- Python: `sum_logprobs` stays as `mx.array` on GPU, accumulated device-side
- Rust: `sum_logprobs` is CPU `f64`, extracted via `.item()` every step

This means Rust's `chosen_logprob` sync is not just expensive — it's **unnecessary**. Python never pulls logprob back to CPU during the decode loop.

### Verification Chain

| Round | Method | Verdict |
|-------|--------|---------|
| 1 | Adversarial testing (6 tests) | Confirmed |
| 2 | Claude Code pixel-level comparison | Confirmed |
| 3 | Original analysis (corrected) | Confirmed |

All three independently agree on the root cause: missing async pipeline overlap + 2x synchronous `.item()` per token.

**Comment by SRjoeee** [comment]
Superseded by #369. The sync count was the wrong abstraction level — the actual root cause is structural serialization from host token round-trips (GPU→host→GPU every token). PR #366 was ineffective.
